### PR TITLE
Upgrade to SpringData MongoDB 3.0 and MongoDB Java Driver 4.0 beta1

### PR DESCRIPTION
**Pull request by christophstrobl**
Upgrade to Spring Data MongoDB 3.0 (Neumann). The update requires MongoDB Driver 4.0 (beta1 at this time). 

Switch from `com.mongodb.MongoClient` and `com.mongodb.MongoClientOptions` to `com.mongodb.client.MongoClient` & `com.mongodb.MongoClientSettings` and share code paths for `MongoClientSettings` between reactive and imperative configuration.

Closes: #19356

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoClientFactorySupport.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+	private void applyHostAndPort(MongoClientSettings.Builder settings) {
+
+		if (isEmbedded()) {
+			settings.applyConnectionString(new ConnectionString("mongodb://localhost:" + getEmbeddedPort()));
+			return;
+		}
+		if (!this.properties.determineUri().equals(MongoProperties.DEFAULT_URI)) {
+			if (hasCustomAddress()) {
+				Assert.state(this.properties.getUri() == null,
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoClientFactorySupport.java]: This is breaking an existing test and I fail to see how that's related to the migration to the new driver. Can you please clarify?

Comment (christophstrobl) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoClientFactorySupport.java]: It seems to be a mismatch that existed between the reactive and imperative implementation. If you remove that `if` statement `ReactiveMongoClientFactoryTests.uriCannotBeSetWithHostPort` will fail which contradicts `MongoPropertiesTests.uriOverridesHostAndPort`.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoAutoConfiguration.java [hunk 2]
```diff
@@ -45,10 +47,13 @@
 public class MongoAutoConfiguration {
 
 	@Bean
-	@ConditionalOnMissingBean(type = { "com.mongodb.MongoClient", "com.mongodb.client.MongoClient" })
-	public MongoClient mongo(MongoProperties properties, ObjectProvider<MongoClientOptions> options,
-			Environment environment) {
-		return new MongoClientFactory(properties, environment).createMongoClient(options.getIfAvailable());
+	@ConditionalOnMissingBean(type = { "com.mongodb.client.MongoClient" })
+	public MongoClient mongo(MongoProperties properties, Environment environment,
+			ObjectProvider<MongoClientSettingsBuilderCustomizer> builderCustomizers,
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoAutoConfiguration.java]: Now that the settings are shared by the two API, we can apply the same customizer consistently which is great. I've removed that in the case of the upgrade though as I'd prefer adding this in an isolated commit.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoClientFactory.java [hunk 3]
```diff
@@ -39,102 +35,23 @@
  * @author Stephane Nicoll
  * @author Nasko Vasilev
  * @author Mark Paluch
+ * @author Christoph Strobl
  * @since 2.0.0
  */
-public class MongoClientFactory {
+public class MongoClientFactory extends MongoClientFactorySupport<MongoClient> {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoClientFactory.java]: While I agree having a shared factory that behaves consistently, I am not sure we are ready to do this just yet. There's quite some history in the original MongoDB auto-configuration and sharing the code with the reactive counter part will change the semantics. I think we should keep the two separated for the time being and investigate what bits are problematic.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/mongo/MongoPropertiesTests.java [hunk 4]
```diff
@@ -55,48 +58,43 @@ void canBindCharArrayPassword() {
 	@Test
 	@SuppressWarnings("deprecation")
 	void allMongoClientOptionsCanBeSet() {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/mongo/MongoPropertiesTests.java]: By restoring the code as it was before in my polish, this test fails. The replicat set is erased by the connection URI. I don't know if that's a new behaviour of the driver or if I messed up the change but I'd appreciate some help there.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/mongo/ReactiveMongoClientFactoryTests.java [hunk 5]
```diff
@@ -117,13 +118,13 @@ void retryWritesIsPropagatedFromUri() {
 	}
 
 	@Test
-	void uriCannotBeSetWithCredentials() {
+	void uriCanBeSetWithCredentials() {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/mongo/ReactiveMongoClientFactoryTests.java]: I don't think we should be making that change as part of the upgrade to the new driver unless this use case is no longer supported.

**Comment by snicoll** [comment]
This broke a smoke test as it looks like we don't connect to the embedded broker properly anymore

**Pull request closed without merging**